[1LP][RFR] Tests tagging on provider side

mshriver opened the pull request:
Rebased from Ruslana Babyuk's original commits in #7320, updated for more recent test patterns.

I would like to leave any remaining test failures to the FA owner to address.

{{ pytest: --long-running -vvvv --use-provider complete cfme/tests/cloud/test_tag_mapping.py cfme/tests/configure/test_tag.py }}

dajoRH commented:
Lint report for commit 79e47ae1add429e8a1f3137b10be7769b60b6a73:

`cfme/tests/cloud/test_tag_mapping.py`:
- :heavy_exclamation_mark: Line 254:9: **W292** *no newline at end of file*

Please, rectify these issues :smirk: .


**No commit flaws detected**.

*CFME QE Bot*

lina-nikiforova commented:
PRT errors:
`AttributeError: 'MapTagsAddView' object has no attribute 'cancel'` and
`AttributeError: 'generator' object has no attribute 'tag_category'`

lina-nikiforova commented on cfme/tests/configure/test_tag.py:
```diff
@@ -50,6 +52,30 @@ def test_tag_crud(category):
+
+    map_tag_entity.delete()
+    view = appliance.browser.create_view(navigator.get_class(map_tags_collection, 'All').VIEW)
+    if not BZ(1510473, forced_streams=['5.9']).blocks:
```
This BZ is closed as a duplicate of 1513507, that one is also closed, `if` is not needed here

lina-nikiforova commented on cfme/tests/configure/test_tag.py:
```diff
@@ -50,6 +52,29 @@ def test_tag_crud(category):
 
 
+def test_map_tagging_crud(appliance, category, soft_assert):
+    """Test Mam tagging grud with flash message assertion"""
```
some typos here :)

mshriver replied:
lol, yeah a few. will clean that up.

lina-nikiforova commented:
@mshriver please check PRT failures

digitronik commented on cfme/tests/cloud/test_tag_mapping.py:
```diff
@@ -41,8 +45,36 @@ def refresh_provider(provider):
+    except ImageNotFoundError:
+        msg = 'Failed looking up template [{}] from CFME on provider: {}'.format(name, provider)
+        logger.exception(msg)
+        pytest.skip(msg)
```
Q: Mike small confusion, why we need to log exception finally we are getting `Skipped` exception from `pytest`?

mshriver replied:
I like to use logger.exception because that will capture the actual exception object and log it, along with my contextual message.

Then pytest.skip with the contextual message, so that is what gets displayed in the artifact report for this failure. The log then contains the complete exception object and traceback.

digitronik commented on cfme/tests/cloud/test_tag_mapping.py:
```diff
@@ -73,32 +105,120 @@ def my_company_tags():
+        # no match / break for select_text
+        if select_text is None:
+            pytest.fail(
+                'Failed to match the entity type [{}] and provider type [{}] in options: [{}]'
```
Optional:
I would use keying in string formatting.

ganeshhubale commented on cfme/tests/cloud/test_tag_mapping.py:
```diff
@@ -41,8 +45,36 @@ def refresh_provider(provider):
+    # Return tuple with random tag_label and tag_value
+    return ('tag_label_{}'.format(fauxfactory.gen_alphanumeric()),
+            'tag_value_{}'.format(fauxfactory.gen_alphanumeric()))
+
```
@mshriver 
Why are we ```tag_components``` is  fixture? We are not even inheriting any other fixture or it is not doing any setup.
Can we define these strings globally?

mshriver replied:
Good point @ganeshhubale.  Didn't want to refactor these commits too much myself, but this is a good change to make. Thanks!

ganeshhubale commented on cfme/tests/cloud/test_tag_mapping.py:
```diff
@@ -24,7 +28,7 @@ def map_tags(appliance, provider, request):
 @pytest.fixture(scope='function')
-def vm(appliance, provider):
+def tagged_vm(appliance, provider):
     # cu-24x7 vm is tagged with test:testing in provider
```
In this fixture, ```appliance``` is [accessed from ```provider``` object.](https://github.com/ManageIQ/integration_tests/pull/8543/files#diff-dcc49450aead0469ee3a75df5dbbf539R34)
So according to me, we can remove ```appliance``` argument.

mshriver commented:
@ganeshhubale @lina-nikiforova @digitronik please take another look, addressed the comments. At this point there are some PRT failures, but I would like the FA owner to address those.

lina-nikiforova commented on cfme/tests/cloud/test_tag_mapping.py:
```diff
@@ -73,32 +104,120 @@ def my_company_tags():
-@pytest.mark.tier(2)
-def test_ec2_tags_mapping():
+@test_requirements.tag
+# TODO: Azure needs tagging support in wrapanapi
```
:+1:

digitronik commented on cfme/tests/cloud/test_tag_mapping.py:
```diff
@@ -24,7 +28,7 @@ def map_tags(appliance, provider, request):
 
 @pytest.fixture(scope='function')
-def vm(appliance, provider):
+def tagged_vm(provider):
```
@mshriver No not related to changes.
I think we need to move it on any available `VM`. No need specific vm for tagging.  
Yup It take time now as UI but I am going to move `VM` `all` method with rest. If you add TODO(ndhandre) it will be great. Just request if possible.

digitronik replied:
as per offline conversation resolving.

digitronik commented on cfme/tests/configure/test_tag.py:
```diff
@@ -284,7 +319,13 @@ def test_assign_and_unassign_tag(self, appliance, tags_mod, provider, services_m
+        try:
+            entity.tags.action.assign(tag)
+        except AttributeError:
+            msg = ('Missing tag attribute in parametrized REST collection {} for entity: {}'
```
Keying please

mshriver replied:
There's only two fields, its clear. I don't think we need to key in every format() call. its useful when there are many fields to fill.

lina-nikiforova commented:
LGTM, I will look further into the failures in a separate PR as it's my FA

izapolsk commented:
since changes look good to me and  @lina-nikiforova promises to handle failed PRT, I'm merging this PRT.
